consensus/src/pending_votes.rs

```diff
@@ -0,0 +1,365 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! PendingVotes store pending votes observed for a fixed epoch and round.
+//! It is meant to be used inside of a RoundState.
+//! The module takes care of creating a QC or a TC
+//! when enough votes (or timeout votes) have been observed.
+//! Votes are automatically dropped when the structure goes out of scope.
+
+use consensus_types::{
+    common::Author, quorum_cert::QuorumCert, timeout_certificate::TimeoutCertificate, vote::Vote,
+};
+use libra_crypto::{hash::CryptoHash, HashValue};
+use libra_logger::prelude::*;
+use libra_types::{
+    ledger_info::LedgerInfoWithSignatures,
+    validator_verifier::{ValidatorVerifier, VerifyError},
+};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt,
+    sync::Arc,
+};
+
+/// Result of the vote processing. The failure case (Verification error) is returned
+/// as the Error part of the result.
+#[derive(Debug, PartialEq)]
+pub enum VoteReceptionResult {
+    /// The vote has been added but QC has not been formed yet. Return the amount of voting power
+    /// the given (proposal, execution) pair.
+    VoteAdded(u64),
+    /// The very same vote message has been processed in past.
+    DuplicateVote,
+    /// The very same author has already voted for another proposal in this round (equivocation).
+    EquivocateVote,
+    /// This block has just been certified after adding the vote.
+    NewQuorumCertificate(Arc<QuorumCert>),
+    /// The vote completes a new TimeoutCertificate
+    NewTimeoutCertificate(Arc<TimeoutCertificate>),
+    /// There might be some issues adding a vote
+    ErrorAddingVote(VerifyError),
+    /// The vote is not for the current round.
+    UnexpectedRound(u64, u64),
+}
+
+/// A PendingVotes structure keep track of votes
+pub struct PendingVotes {
+    /// Maps LedgerInfo digest to associated signatures (contained in a partial LedgerInfoWithSignatures).
+    /// This might keep multiple LedgerInfos (for the current round) to tolerate non-determinism in execution.
```

Comment:
this is not to tolerate non-determinism, a round can have a good proposal and a NIL proposal

Reply:
yeah actually execution should always be deterministic oO (I just copied that comment), IIUC it is possible that if the proposer creates different proposals, we end up receiving timeout votes that carry different proposals as well right? In this case we would save them under different li_digest_to_votes

Reply:
> then we can end up receiving timeout votes that carry different proposals as well right?

yup

Reply:
ok I'll modify the comment according to this!

Reply:
which makes me think, we should log the receiving of different non-NIL proposals for the same round